Add workaround for compiler assertion failure in Promise.flatten_next (#3991)
ponylang/ponyc: .release-notes/3991.md

## Add workaround for compiler assertion failure in Promise.flatten_next

> Under certain conditions a specific usage of `Promise.flatten_next` and `Promise.next` could
> trigger a compiler assertion, whose root-cause is still to be investigated. 

Reviewer: ```suggestion
Under certain conditions a specific usage of `Promise.flatten_next` and `Promise.next` could trigger a compiler assertion, whose root-cause is still under investigation. 
```

> This change avoid the code-path triggering the assertion entirely, and is thus only a workaround, not a full solution, but it increases stability of the stdlib and the promises package.

Reviewer: ```suggestion
The workaround avoids the code-path that triggers the assertion. We still need a full solution.
```
